docs/literate/src/files/differentiable_programming_literate.jl

```diff
@@ -0,0 +1,409 @@
+#src # Differentiable programming
+using Test #src
+
+# [Julia and its ecosystem provide some tools for differentiable programming](https://sinews.siam.org/Details-Page/scientific-machine-learning-how-julia-employs-differentiable-programming-to-do-it-best).
+# Trixi.jl is designed to be flexible, extendable, and composable with Julia's growing ecosystem for
+# scientific computing and machine learning. Thus, the ultimate goal is to have fast implementations
+# that allow automatic differentiation (AD) without too much hassle for users. If some parts do not
+# meet these requirements, please feel free to open an issue or propose a fix in a PR.
+
+# In the following, we will walk through some examples demonstrating how to differentiate through
+# Trixi.jl.
+
+
+# ## Forward mode automatic differentiation
+
+# Trixi integrates well with [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl)
+# for forward mode AD.
+
+
+# ### Computing the Jacobian
+# The high-level interface to compute the Jacobian this way is [`jacobian_ad_forward`](@trixi-docs:reference-trixi/#Trixi.jacobian_ad_forward-Tuple{Trixi.AbstractSemidiscretization}).
+
+using Trixi, LinearAlgebra, Plots
+#-
+equations = CompressibleEulerEquations2D(1.4);
+
+solver = DGSEM(3, flux_central);
+
+mesh = TreeMesh((-1.0, -1.0), (1.0, 1.0), initial_refinement_level=2, n_cells_max=10^5);
+
+semi = SemidiscretizationHyperbolic(mesh, equations, initial_condition_density_wave, solver);
+
+J = jacobian_ad_forward(semi);
+
+size(J)
+@test size(J) == (1024, 1024) #src
+#-
+λ = eigvals(J);
+
+scatter(real.(λ), imag.(λ));
+
+3.0e-10 < maximum(real, λ) / maximum(abs, λ) < 8.0e-10
+@test (3.0e-10 < maximum(real, λ) / maximum(abs, λ) < 8.0e-10) == true #src
+#-
+1.0e-7 < maximum(real, λ) < 5.0e-7
+@test (1.0e-7 < maximum(real, λ) < 5.0e-7) == true #src
```

Comment:
In the static version this would look like
```jldoctest euler_eigenvalues
julia> using Trixi, LinearAlgebra, Plots

julia> equations = CompressibleEulerEquations2D(1.4);

julia> solver = DGSEM(3, flux_central);

julia> mesh = TreeMesh((-1.0, -1.0), (1.0, 1.0), initial_refinement_level=2, n_cells_max=10^5);

julia> semi = SemidiscretizationHyperbolic(mesh, equations, initial_condition_density_wave, solver);

julia> J = jacobian_ad_forward(semi);

julia> size(J)
(1024, 1024)

julia> λ = eigvals(J);

julia> scatter(real.(λ), imag.(λ));

julia> 3.0e-10 < maximum(real, λ) / maximum(abs, λ) < 8.0e-10
true

julia> 1.0e-7 < maximum(real, λ) < 5.0e-7
true
```